framework/src/interfacekernels/ADInterfaceKernel.C

```diff
@@ -0,0 +1,317 @@
+  else
+    prepareVectorTagNeighbor(_assembly, _neighbor_var.number());
+
+  for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+    for (_i = 0; _i < test_space.size(); _i++)
+      _local_re(_i) += raw_value(_ad_JxW[_qp] * _ad_coord[_qp] * computeQpResidual(type));
+
+  accumulateTaggedLocalResidual();
+}
+
+template <typename T>
+void
+ADInterfaceKernelTempl<T>::computeResidual()
+{
+  // in the gmsh mesh format (at least in the version 2 format) the "sideset" physical entities are
+  // associated only with the lower-dimensional geometric entity that is the boundary between two
+  // higher-dimensional element faces. It does not have a sidedness to it like the exodus format
+  // does. Consequently we may naively try to execute an interface kernel twice, one time where _var
+  // has dofs on _current_elem *AND* _neighbor_var has dofs on _neighbor_elem, and the other time
+  // where _var has dofs on _neighbor_elem and _neighbor_var has dofs on _current_elem. We only want
+  // to execute in the former case. In the future we should remove this and add some kind of "block"
+  // awareness to interface kernels to avoid all the unnecessary reinit that happens before we hit
+  // this return
+  if (!_var.activeOnSubdomain(_current_elem->subdomain_id()) ||
+      !_neighbor_var.activeOnSubdomain(_neighbor_elem->subdomain_id()))
+    return;
+
+  // Compute the residual for this element
+  computeElemNeighResidual(Moose::Element);
+
+  // Compute the residual for the neighbor
+  computeElemNeighResidual(Moose::Neighbor);
+}
+
+template <typename T>
+void
+ADInterfaceKernelTempl<T>::computeElemNeighJacobian(Moose::DGJacobianType type)
+{
+  const ADTemplateVariableTestValue<T> & test_space =
+      (type == Moose::ElementElement || type == Moose::ElementNeighbor) ? _test : _test_neighbor;
+  const ADTemplateVariableTestValue<T> & loc_phi =
+      (type == Moose::ElementElement || type == Moose::NeighborElement) ? _phi : _phi_neighbor;
+
+  unsigned int ivar, jvar;
+
+  switch (type)
+  {
+    case Moose::ElementElement:
+      ivar = jvar = _var.number();
+      resid_type = Moose::Element;
+      break;
+    case Moose::ElementNeighbor:
+      ivar = _var.number(), jvar = _neighbor_var.number();
+      resid_type = Moose::Element;
+      break;
+    case Moose::NeighborElement:
+      ivar = _neighbor_var.number(), jvar = _var.number();
+      resid_type = Moose::Neighbor;
+      break;
+    case Moose::NeighborNeighbor:
+      ivar = _neighbor_var.number(), jvar = _neighbor_var.number();
+      resid_type = Moose::Neighbor;
+      break;
+    default:
+      mooseError("Unknown DGJacobianType ", type);
+  }
+
+  if (type == Moose::ElementElement)
+    prepareMatrixTag(_assembly, ivar, jvar);
+  else
+    prepareMatrixTagNeighbor(_assembly, ivar, jvar, type);
+
+  std::size_t ad_offset;
+  if (type == Moose::ElementElement || type == Moose::NeighborElement)
+    ad_offset = jvar * _sys.getMaxVarNDofsPerElem();
```

Comment:
@cticenhour I believe the key change was here. I think you mostly copy-pasted this from `ADDGKernel` maybe where this isn't a concern...but previously I believe you had `_var.number()`. So we were getting the jacobian wrong when computing the Jacobians with respect to the neighbor residual because we were always using the "elemental" variable. So the change was to go from `_var.number()` -> `jvar`.

Reply:
Ah you're right - I originally had `ivar` here. I actually didn't look at `ADDGKernel`, but a combo of `ADIntegratedBC` and `ADKernel` and then built something up from what I saw there and my understanding of `InterfaceKernel`. My core misunderstanding was the concept of an `ad_offset` in the first place, it seems. Could you explain that a bit more?

Reply:
I have to remind myself about this sometimes because I forget...

Let's just think about continuous finite elements for a second. We use a variable major numbering scheme, such that each variables indices are in a contiguous block. Let's imagine we have two variables, u and v, and we're on a QUAD4. The AD indices will be ordered like this:

u0, u1, u2, u3, v0, v1, v2, v3

`_sys.getMaxVarNDofsPerElem` should return for a QUAD4: 4. For a QUAD9, 9. HEX27, 27. Etc. For CFEM the offset will be simply that element number times the var number. So for u on a QUAD4: 4 * 0 = 0. For v: 4 * 1. So u indices start at index 0, v indices start at index 4.

With DFEM or interface kernels it's a little more complicated. We essentially already have an indices block that is num_vars * num_dofs_per_elem long, so in our two var, QUAD4 example: 4 * 2 = 8. So what we do is that if we are on a neighbor element, we do an *additional* offset by num_vars * num_dofs_per_elem. So now our derivative indices are ordered like this:

u0, u1, u2, u3, v0, v1, v2, v3, u0_neighbor, u1_neighbor, u2_neighbor, u3_neighbor, v0_neighbor, v1_neighbor, v2_neighbor, v3_neighbor

Does that make sense?

Reply:
Very clear - thanks! :)

Reply:
> Let's just think about continuous finite elements for a second. We use a variable major numbering scheme, such that each variables indices are in a contiguous block. Let's imagine we have two variables, u and v, and we're on a QUAD4. The AD indices will be ordered like this:
> 
> u0, u1, u2, u3, v0, v1, v2, v3
> 
> `_sys.getMaxVarNDofsPerElem` should return for a QUAD4: 4. For a QUAD9, 9. HEX27, 27. Etc. For CFEM the offset will be simply that element number times the var number. So for u on a QUAD4: 4 * 0 = 0. For v: 4 * 1. So u indices start at index 0, v indices start at index 4.
> 
> With DFEM or interface kernels it's a little more complicated. We essentially already have an indices block that is num_vars * num_dofs_per_elem long, so in our two var, QUAD4 example: 4 * 2 = 8. So what we do is that if we are on a neighbor element, we do an _additional_ offset by num_vars * num_dofs_per_elem. So now our derivative indices are ordered like this:
> 
> u0, u1, u2, u3, v0, v1, v2, v3, u0_neighbor, u1_neighbor, u2_neighbor, u3_neighbor, v0_neighbor, v1_neighbor, v2_neighbor, v3_neighbor

I would really appreciate it if you could add these sentences as comments into code.